File: app/src/main/java/com/worldventures/dreamtrips/social/ui/tripsimages/service/command/FetchLocationFromExifCommand.java
package com.worldventures.dreamtrips.social.ui.tripsimages.service.command;


import android.media.ExifInterface;

import com.worldventures.dreamtrips.modules.trips.model.Location;

import java.io.IOException;

import io.techery.janet.Command;
import io.techery.janet.command.annotations.CommandAction;
import rx.Observable;

@CommandAction
public class FetchLocationFromExifCommand extends Command<Location> {

   private final String filePath;

   public FetchLocationFromExifCommand(String filePath) {
      this.filePath = filePath;
   }

   @Override
   protected void run(CommandCallback<Location> callback) throws Throwable {
      Observable.just(filePath)
            .flatMap(path -> {
               ExifInterface exifInterface = null;
               try {
                  exifInterface = new ExifInterface(path);
               } catch (IOException e) {
                  return Observable.error(e);
               }
               float latLng[] = new float[2];
               if (exifInterface.getLatLong(latLng)) {
                  return Observable.just(new Location(latLng[0], latLng[1]));
               }
               return Observable.just(new Location());
            })
            .subscribe(callback::onSuccess, callback::onFail);
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/ui/tripsimages/service/command/FetchLocationFromExifCommand.kt
package com.worldventures.dreamtrips.social.ui.tripsimages.service.command

import android.media.ExifInterface
import com.worldventures.dreamtrips.modules.trips.model.Location
import io.techery.janet.Command
import io.techery.janet.command.annotations.CommandAction
import rx.Observable

@CommandAction
class FetchLocationFromExifCommand(private val filePath: String) : Command<Location>() {

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<Location>) {
      Observable.just(filePath)
            .flatMap { path ->
               val exifInterface = ExifInterface(path)
               val latLng = FloatArray(2)
               if (exifInterface.getLatLong(latLng)) {
                  Observable.just(Location(latLng[0].toDouble(), latLng[1].toDouble()))
               } else {
                  Observable.just(Location())
               }
            }
            .subscribe(callback::onSuccess, callback::onFail)
   }
}
